zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -771,11 +737,7 @@ pub(crate) mod tests {
                 None,
                 None
             ),
-            Err(data_api::error::Error::InsufficientFunds {
-                available,
-                required
-            })
-            if available == Amount::zero() && required == Amount::from_u64(10001).unwrap()
+            Err(data_api::error::Error::ScanRequired)
```

Comment:
This change suggests that we currently have a bug in `init_blocks_table` (either in the code or documentation).

Before SbS the `blocks` table was the authoritative source of truth for the wallet's view of the chain tip, so by definition `init_blocks_table` was setting the wallet's initial view of the chain tip, and `InsufficientFunds` is the correct error.

With SbS, we now have two scenarios:
- We are doing linear scanning, in which case the `blocks` table's maximum block height will be equal to the highest scanned height in the scan queue, so either can be used.
- We are doing SbS, in which case we expect the scan queue to have heights greater than the `blocks` table as a result of calling `update_chain_tip`, and to be the authoritative source of truth for the wallet's view of the chain tip.

In the SbS case, `ScanRequired` makes sense as an error that can occur when there is a gap between the `blocks` table and the scan queue (or equivalently, when there are non-`Scanned` ranges in the scan queue above the maximum height in the `blocks` table). But if all ranges are scanned, then we should see `InsufficientFunds`, and in the test above where we never call `update_chain_tip`, the _only_ source of information is the `blocks` table, and what it was initialised to; hence in the wallet's view it is fully-synchronised, and should return `InsufficientFunds`.

The question is what to do about this. I see two mutually-exclusive pathways:
- Alter `init_blocks_table` to ensure we preserve its current semantics: that it fully-initialises a new wallet's view of the chain tip to be at the given block height. Existing wallets continue to call it without changes, and then those wallets that additionally use SbS will naturally call `update_chain_tip` to move the view forward.
- Keep the implementation of `init_blocks_table` as-is, and change its documentation to indicate that it is now an error to _not_ call `update_chain_tip` afterwards. This should be coupled with pervasive error condition checking to ensure that wallets don't miss the fact that they need to change their code. That could reuse the `ScanRequired` error, but it feels to me like this should be a separate `NotFullyInitialised` error or similar (because the fix is _not_ to call `suggest_scan_ranges` and scan them - that won't fix the problem - but to call `update_chain_tip`).

Reply:
> We are doing linear scanning, in which case the blocks table's maximum block height will be equal to the highest scanned height in the scan queue, so either can be used.

This is incorrect. When we call `update_chain_tip` we will create a new scan entry with the range `previous_max_scanned+1..new_chain_tip` and these two values will not be in agreement. Linear scanning is still implemented in terms of `suggest_scan_ranges`. 

I don't like the inconsistency of having `init_blocks_table` add metadata for a block, when we haven't scanned that block, while in all other cases we add block metadata _after_ having scanned the block. Instead, one possibility is that `init_blocks_table` (which should probably be renamed) does two things: first, it adds the note commitment tree frontier that it is provided to the note commitment tree, making it possible to scan from that point forward, and then secondly it should create the 1-block scan range that starts with that block. 

This would also fix an issue that the wallet developers have been noticing that affects testnet wallets, which is that since there are so few Sapling notes on testnet, the block range that must be scanned in order to fill out the current-tip shard is enormous. What we should do is explicitly add a method to set the wallet birthday that takes the birthday height and the note commitment tree frontier as of that height; we should save the wallet birthday explicitly (perhaps in a different table) and then ensure that the view of unscanned ranges that is created in #894 uses this birthday height as the lower bound for the scan ranges, rather than Sapling activation.

Reply:
> This is incorrect. When we call `update_chain_tip` we will create a new scan entry with the range `previous_max_scanned+1..new_chain_tip` and these two values will not be in agreement. Linear scanning is still implemented in terms of `suggest_scan_ranges`.

Hmm, I did not pick this up across the multiple PRs that implemented SbS. My understanding from our previous discussions was that they were being implemented such that an existing wallet doing linear scanning could keep doing what they were before in terms of their scanning life cycle (after updating for the breaking changes to the APIs they were using, per the changelog), and it would continue to work.

The changelog entries for this release definitely need to explain this semantic change (and the fact that pure linear scanning has been removed, replaced by "linear scanning SbS"). If I were to follow the instructions of the changelog entries as currently written, but go no further to implement SbS, my wallet would be broken. We also need a test for the error(s) that these wallets would see if they did not change their scanning life cycle to be "linear scanning SbS".

> I don't like the inconsistency of having `init_blocks_table` add metadata for a block, when we haven't scanned that block, while in all other cases we add block metadata _after_ having scanned the block.

It wasn't inconsistent with my view of how SbS was being integrated (per above). The definition of `init_blocks_table` is that the block you are initializing with _is_ fully scanned, because it has nothing of interest to the wallet (there are no notes in that block or any prior one, and you're providing a Sapling frontier that means this and prior blocks do not need to be scanned for `ShardTree` purposes). So it forms a valid fully-scanned start point for both old linear scanning, and new SbS.

I'm glad we've uncovered this discrepancy in how you and I were viewing the SbS changes; I had previously said that we needed to be very careful to look out for breaking changes to the assumptions inherent in the `blocks` table, and this is another such assumption-break. I'm worried how much extra work this might be to resolve though, compared to preserving the old linear scanning workflow (though maybe this far into the SbS PRs it's now more work to do the latter, IDK). Let's chat about this offline.

Reply:
I don't see a use case for "update to this new version of the SDK, which has major breaking changes that I need to address in a bunch of places, but don't update to the new scan logic." If someone wants to keep with linear scanning, they just don't have to update.

Reply:
The problem is that the fact that the crate releases completely breaks the old scanning logic _isn't documented anywhere_. So users can rightfully believe that they have done everything they need to, based on the current changelog (and even after the missing changelog entries from previous PRs are added as tracked by #860), their code would compile, and would likely work for some test scenarios - and then they deploy a broken wallet update.

If the old scanning logic is going to remain broken, then (in addition to the various code and test changes we still need to make due to breaking more assumptions about the `blocks` table) the crate releases are blocked on documenting this in a way that people upgrading can easily identify and follow.

Reply:
In an offline meeting, @nuttycom and I decided that what we should do is replace `init_blocks_table` with a dedicated initialization method that handles the `blocks` table and scan queue, and sets up any relevant other state. Kris thinks it makes sense for this to also be tied to account initialization, but I'm not yet convinced and am waiting on the PR to review 🙂

Reply:
This issue is now tracked as #901